From 58c0fb173ff5b4a8b0456529db0766f4b9ddbece Mon Sep 17 00:00:00 2001
From: Omri Itach <omrii@marvell.com>
Date: Mon, 24 Jun 2013 16:19:21 +0300
Subject: [PATCH 0733/1825] fix: alp: Fixed mac speed configuration (according
 to Eth complex settings)

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit a592d83291851ebb8b867eba30ece7930229ddc7

	- Added function mvBoardMacSpeedSet to update Mac speed after scanning Eth complex configuration

Change-Id: I2f84c4c3da2bdd40fae426597833cf764643813e
Signed-off-by: Omri Itach <omrii@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/2344
Reviewed-by: Dmitri Epshtein <dima@marvell.com>
Reviewed-by: Star_Automation <star@marvell.com>
Reviewed-by: Jonatan Farhadian <yonif@marvell.com>
Reviewed-by: Eran Ben-Avi <benavi@marvell.com>
Tested-by: Eran Ben-Avi <benavi@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 .../avanta_lp_family/boardEnv/mvBoardEnvLib.c      |   53 ++++++++++++++++---
 .../avanta_lp_family/boardEnv/mvBoardEnvLib.h      |    1 +
 2 files changed, 45 insertions(+), 9 deletions(-)

diff --git a/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.c b/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.c
index 39f02f2..e68aa77 100644
--- a/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.c
+++ b/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.c
@@ -389,6 +389,31 @@ MV_BOARD_MAC_SPEED mvBoardMacSpeedGet(MV_U32 ethPortNum)
 }
 
 /*******************************************************************************
+* mvBoardMacSpeedSet - Set the Mac speed
+*
+* DESCRIPTION:
+*       This routine Sets the Mac speed of a given ethernet port.
+*
+* INPUT:
+*       ethPortNum - Ethernet port number.
+*       macSpeed   - Requested MAC speed
+*
+* OUTPUT:
+*       None.
+*
+* RETURN:
+*       None
+*
+*******************************************************************************/
+MV_VOID mvBoardMacSpeedSet(MV_U32 ethPortNum, MV_BOARD_MAC_SPEED macSpeed)
+{
+	if (ethPortNum >= board->numBoardMacInfo)
+		mvOsPrintf("%s: Error: wrong eth port (%d)\n", __func__, ethPortNum);
+
+	board->pBoardMacInfo[ethPortNum].boardMacSpeed = macSpeed;
+}
+
+/*******************************************************************************
 * mvBoardIsPortLoopback -
 *
 * DESCRIPTION:
@@ -837,7 +862,8 @@ MV_VOID mvBoardMppTypeSet(MV_U32 mppGroupNum, MV_U32 groupType)
 *******************************************************************************/
 MV_VOID mvBoardInfoUpdate(MV_VOID)
 {
-	MV_U32 slicDev, ethComplex;
+	MV_U32 smiAddress, slicDev, ethComplex;
+	MV_BOARD_MAC_SPEED macSpeed = BOARD_MAC_SPEED_AUTO; /*if Mac is not connected to switch, auto-negotiate speed*/
 
 	/* Update Ethernet complex according to board configuration */
 	mvBoardEthComplexInfoUpdate();
@@ -845,18 +871,27 @@ MV_VOID mvBoardInfoUpdate(MV_VOID)
 	/* Update SMI phy address for MAC0/1 */
 	ethComplex = mvBoardEthComplexConfigGet();
 	if (ethComplex & MV_ETHCOMP_GE_MAC0_2_GE_PHY_P0)
-		mvBoardPhyAddrSet(0, 0x0);
+		smiAddress = 0x0;
 	else if (ethComplex & MV_ETHCOMP_GE_MAC0_2_RGMII0)
-		mvBoardPhyAddrSet(0, 0x4);
-	else
-		mvBoardPhyAddrSet(0, -1); /* no SMI address if connected to switch */
+		smiAddress = 0x4;
+	else {
+		smiAddress = -1; /* no SMI address if connected to switch */
+		macSpeed = BOARD_MAC_SPEED_1000M;
+	}
+	mvBoardPhyAddrSet(0, smiAddress);
+	mvBoardMacSpeedSet(0, macSpeed);
 
+	macSpeed = BOARD_MAC_SPEED_AUTO; /*if Mac is not connected to switch, auto-negotiate speed*/
 	if (ethComplex & MV_ETHCOMP_GE_MAC1_2_GE_PHY_P3)
-		mvBoardPhyAddrSet(1, 0x3);
+		smiAddress = 0x3;
 	else if (ethComplex & MV_ETHCOMP_GE_MAC1_2_RGMII1)
-		mvBoardPhyAddrSet(1, 0x1);
-	else
-		mvBoardPhyAddrSet(1, -1); /* no SMI address if connected to switch */
+		smiAddress = 0x1;
+	else {
+		smiAddress = -1; /* no SMI address if connected to switch */
+		macSpeed = BOARD_MAC_SPEED_1000M;
+	}
+	mvBoardPhyAddrSet(1, smiAddress);
+	mvBoardMacSpeedSet(1, macSpeed);
 
 	/* Update MPP group types and values according to board configuration */
 	mvBoardMppIdUpdate();
diff --git a/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.h b/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.h
index 270aed7..31aff78 100644
--- a/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.h
+++ b/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.h
@@ -455,6 +455,7 @@ MV_BOOL mvBoardIsQsgmiiModuleConnected(void);
 MV_32 mvBoardGePhySwitchPortGet(void);
 MV_32 mvBoardRgmiiASwitchPortGet(void);
 MV_BOARD_MAC_SPEED mvBoardMacSpeedGet(MV_U32 ethPortNum);
+MV_VOID mvBoardMacSpeedSet(MV_U32 ethPortNum, MV_BOARD_MAC_SPEED macSpeed);
 MV_U32 mvBoardSwitchCpuPortGet(MV_U32 switchIdx);
 MV_32 mvBoardSwitchIrqGet(MV_VOID);
 MV_32 mvBoardSwitchConnectedPortGet(MV_U32 ethPort);
-- 
1.7.5.4

